Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs

```diff
@@ -174,118 +169,160 @@ public IConnection Acquire(AccessMode mode)
+
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        if (!_availableConnections.TryTake(out connection, 0, cancellationToken))
                         {
-                            break;
+                            do
+                            {
+                                if (!IsConnectionPoolFull())
+                                {
+                                    connection = CreateNewPooledConnection();
+                                    break;
+                                }
+
+                                if (_availableConnections.TryTake(out connection, 0, cancellationToken))
+                                {
+                                    break;
+                                }
+                            } while (!cancellationToken.IsCancellationRequested);
+
+                            if (connection == null)
+                            {
+                                throw new ClientException(
+                                    $"Failed to obtain a connection from pool within {_connAcquisitionTimeout}");
+                            }
                         }
-                    }
-                    while (connAcquisitionTimer.ElapsedMilliseconds < _connAcquisitionTimeout.TotalMilliseconds);
-                    connAcquisitionTimer.Stop();
 
-                    if (connection == null)
-                    {
-                        throw new ClientException($"Failed to obtain a connection from pool within {_connAcquisitionTimeout}");
+                        if (!_connectionValidator.IsValid(connection))
+                        {
+                            DestoryConnection(connection);
+                        }
+                        else
+                        {
+                            break;;
+                        }
                     }
-                }
 
-                if (!_connectionValidator.IsValid(connection))
-                {
-                    DestoryConnection(connection);
-                    return Acquire();
-                }
-
-                _inUseConnections.TryAdd(connection);
-                if (_disposeCalled)
-                {
-                    if (_inUseConnections.TryRemove(connection))
+                    _inUseConnections.TryAdd(connection);
+                    if (_disposeCalled)
                     {
-                        DestoryConnection(connection);
+                        if (_inUseConnections.TryRemove(connection))
+                        {
+                            DestoryConnection(connection);
+                        }
+
+                        ThrowObjectDisposedException();
                     }
-                    ThrowObjectDisposedException();
-                }
 
-                return connection;
-            });
+                    return connection;
+                });
+            }
+            catch (OperationCanceledException ex)
```

Comment:
I am thinking that this `try-catch` should be inside `return TryExecute`. TryExecute is only used to log errors before throwing the error to user. I would think this `OperationCanceledException` is less interesting to an end user than `ClientException`